### PR TITLE
kubelet: add support for reverse-proxying endpoints from config

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind feature
#### What this PR does / why we need it:
This is an alternative to https://github.com/kubernetes/kubernetes/pull/111355, with the same motivation. The main differentiating factor is where the endpoints are specified from. This specifies from the config, which could be useful for a plugin architecture to relay the metrics endpoint. 

Below is a copy of the description of that PR, for the lazy among us:
Currently, the Kubelet relays cAdvisor's Prometheus metrics that are collected for containers. One piece of this relaying is these metrics are propagated through the kubelet's HTTP endpoint. This endpoint has a number of quality of life aspects: there is authentication, and proxying by the kube API. As we move to the CRI implementation broadcasting these metrics, it would be best if those QOL aspects aren't lost.

Do this by implementing a config field ProxiedEndpoints that an admin can point to the URL of the CRI implementation (or a plugin) to do this relaying of metrics 

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Add Kubelet support for ProxiedEndpoints configuration. This allows an admin to specify endpoints for the kubelet to reverse proxy. This could be useful in proxying metrics from an entity that isn't cAdvisor (like a CRI plugin) 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/2371
```

### CODE REVIEW

File: api/api-rules/violation_exceptions.list [hunk 1]
```diff
@@ -360,6 +360,7 @@ API rule violation: list_type_missing,k8s.io/kubelet/config/v1beta1,CredentialPr
 API rule violation: list_type_missing,k8s.io/kubelet/config/v1beta1,KubeletConfiguration,AllowedUnsafeSysctls
 API rule violation: list_type_missing,k8s.io/kubelet/config/v1beta1,KubeletConfiguration,ClusterDNS
 API rule violation: list_type_missing,k8s.io/kubelet/config/v1beta1,KubeletConfiguration,EnforceNodeAllocatable
+API rule violation: list_type_missing,k8s.io/kubelet/config/v1beta1,KubeletConfiguration,ProxiedEndpoints
```
Comment (haircommander) [inline api/api-rules/violation_exceptions.list:363]: I used [this](https://github.com/kubernetes/kubernetes/pull/102915/commits/d82f6069708ff25ecff48464bee82bc570b55641#r740407436) comment to justify this change

**Comment by aojea** [comment]
challenging question about security, can this be abused and present a security risk?
we are adding a proxy that can redirect to any place via the kubelet config

**Comment by haircommander** [comment]
> challenging question about security, can this be abused and present a security risk? we are adding a proxy that can redirect to any place via the kubelet config

I don't think this exposes any risks that aren't exposed with kubelet config access. If a malicious user can access the kubelet config, they can schedule any process (reroute the CRI socket used to a malicious one that schedules bad pods) and a host of other things I haven't thought of.

**Comment by liggitt** [comment]
> /label api-review
> 
> (there's no corresponding KEP update yet. We're still deciding on the implementation)

Sorry, not sure how my query of PRs with the label missed this PR. I can't tell if the comment about a KEP update is resolved at this point

**Comment by haircommander** [comment]
> > /label api-review
> > (there's no corresponding KEP update yet. We're still deciding on the implementation)
> 
> Sorry, not sure how my query of PRs with the label missed this PR. I can't tell if the comment about a KEP update is resolved at this point... were there still questions about the implementation approach?

@bobbypage and I have yet to decide on an approach. It's likely at this point we'll slip 1.25, unless we get an API review and a node review in the next two days (which I don't expect).

**Comment by liggitt** [comment]
> @bobbypage and I have yet to decide on an approach. It's likely at this point we'll slip 1.25, unless we get an API review and a node review in the next two days (which I don't expect).

That's helpful context... in general, the review of the supporting API comes after the owning sig has an functional approach they are satisfied with (pinning that down often changes the shape of the information needed in the config or API).

I'll drop api-review from this for now, once there's an approach settled on, requeue it for config review

/remove-label api-review

**Comment by haircommander** [comment]
we've opted for a different approach https://github.com/kubernetes/enhancements/pull/3559

**Pull request closed without merging**